Improved error handling for test command

treydock opened the pull request:
## Related issue

<!--
Please link the GitHub issue this pull request resolves in the format of `#1234`. If you discussed this change
with a maintainer, please mention her/him using the `@` syntax (e.g. `@JimBugwadia`).

If this change neither resolves an existing issue nor has sign-off from one of the maintainers, there is a
chance substantial changes will be requested or that the changes will be rejected.

You can discuss changes with maintainers in the [Kyverno Slack Channel](https://kubernetes.slack.com/).
-->
#1865

## What type of PR is this

<!--

> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading white spaces from that line:
>
> /kind api-change
> /kind cleanup
> /kind design
> /kind documentation
> /kind failing-test
> /kind feature
-->
> /kind bug

## Proposed Changes

<!--
Describe the big picture of your changes here to communicate to the maintainers why we should accept this pull request. 

***NOTE***: If this PR results in new or altered behavior which is user facing, you **MUST** read and follow the steps outlined in the [PR documentation guide](pr_documentation.md) and add Proof Manifests as defined below.
-->
Without this change, the testing CLI doesn't actually work when using a directory. First main issue is the number of found YAML files was never actually made available so always seen as 0.  Once that issue is resolved the errors generated for both https:// and directory tests were never actually added to `errors` slice so they were never actually printed. I found this out using malformed `test.yaml` and didn't know my issues were related to incorrect YAML until made errors available.

### Proof Manifests

I added manifests from here: https://kyverno.io/docs/kyverno-cli/#test to be executed during testing.

## Checklist

<!--
Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of
them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code.
-->

- [x] I have read the [contributing guidelines](https://github.com/kyverno/kyverno/blob/main/CONTRIBUTING.md).
- [x] I have added tests that prove my fix is effective or that my feature works.
- [] My PR contains new or altered behavior to Kyverno and
  - [] I have added or changed [the documentation](https://github.com/kyverno/website) myself in an existing PR and the link is:
  <!-- Uncomment to link to the PR -->
  <!-- https://github.com/kyverno/website/pull/123 -->
  - [] I have raised an issue in [kyverno/website](https://github.com/kyverno/website) to track the doc update and the link is:
  <!-- Uncomment to link to the issue -->
  <!-- https://github.com/kyverno/website/issues/1 -->
  - [] I have read the [PR documentation guide](pr_documentation.md) and followed the process including adding proof manifests to this PR.

## Further Comments

<!--
If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution
you did and what alternatives you considered, etc...
-->

I would note that I think the test for https://github.com/kyverno/policies/main is actually failing but not actually showing up:

```
make run_testcmd_policy
...
Executing drop-all-capabilities...
applying 1 policy to 1 resource... 
│───│───────────────────────────────────────────────────────────────────│────────│
│ # │ TEST                                                              │ RESULT │
│───│───────────────────────────────────────────────────────────────────│────────│
│ 1 │ add-capabilities with drop-all-capabilities/check-init-containers │ Fail   │
│───│───────────────────────────────────────────────────────────────────│────────│

```

treydock commented:
https://github.com/kyverno/policies/issues/54 solves test that was showing as failed but not failing.  I noticed from code that if a test is defined for a rule, and that rule and test don't match, it would show as `Fail` because that's the default. I changed default to `Not found` and if something fails then it actually prints `Fail`. I'm not sure if something like `Not found` should also exit with 1 or continue to exit with 0.

vyankyGH commented on pkg/engine/jsonContext.go:
```diff
@@ -27,6 +27,9 @@ func LoadContext(logger logr.Logger, contextEntries []kyverno.ContextEntry, resC
 	if store.GetMock() {
 		rule := store.GetPolicyRuleFromContext(policyName, ruleName)
+		if len(rule.Values) == 0 {
+			return errors.New(fmt.Sprintf("No values found for policy %s rule %s", policyName, ruleName))
```
@treydock can you please fix e22-test action.

treydock replied:
What is it needing to be fixed? On my end I see all Github actions as successful.  I would like to add some e2e tests using policies with variables which is what part of this pull requests fixes.

treydock replied:
I see the error now, I had to view the pull request diff.  I fixed this issue.

treydock commented:
Added commits to fix handling of parsing JSON arrays in test variables and also defining multiple policies to test in same set of files.  I added e2e tests to validate all this behavior.

treydock commented:
There is a problem with 00b8da9219f8f7446177ec42c07879489f63eb37 where tests without variables will fail it seems like.  I will have to adjust that logic I think, trying to see why the e2e tests didn't pick that up

vyankyGH commented:
> There is a problem with [00b8da9](https://github.com/kyverno/kyverno/commit/00b8da9219f8f7446177ec42c07879489f63eb37) where tests without variables will fail it seems like. I will have to adjust that logic I think, trying to see why the e2e tests didn't pick that up

look like this commit does not belong to any branch

treydock commented:
Just pushed a commit to ensure the logic to check policy names matches is only used when variables are present since that's when multiple policies defined in a single variables file was causing issues.  The test was reporting `Not found` which exits 0.  Once this is merged I will try to open new pull request to maybe improve exit codes around tests where nothing is found as right now a policy test that is incorrectly configured is reported the same as a policy test where the resources don't match the policy selector.

NoSkillGirl commented:
looks good to me

NoSkillGirl reviewed (approved):
looks good to me
